lib/compat/wordpress-7.0/blocks.php

```diff
@@ -0,0 +1,107 @@
+<?php
+/**
+ * Compatibility shims for block APIs for WordPress 7.0.
+ *
+ * @package gutenberg
+ */
+
+if ( ! function_exists( 'gutenberg_resolve_pattern_blocks' ) ) {
+    /**
+     * Replaces patterns in a block tree with their content.
+     *
+     * @since 6.6.0
+     * @since 7.0.0 Adds metadata to attributes of single-pattern container blocks.
+     *
+     * @param array $blocks An array blocks.
+     *
+     * @return array An array of blocks with patterns replaced by their content.
+     */
+    function gutenberg_resolve_pattern_blocks( $blocks ) {
+        static $inner_content;
+        // Keep track of seen references to avoid infinite loops.
+        static $seen_refs = array();
+        $i                = 0;
+        while ( $i < count( $blocks ) ) {
+            if ( 'core/pattern' === $blocks[ $i ]['blockName'] ) {
+                $attrs = $blocks[ $i ]['attrs'];
+
+                if ( empty( $attrs['slug'] ) ) {
+                    ++$i;
+                    continue;
+                }
+
+                $slug = $attrs['slug'];
+
+                if ( isset( $seen_refs[ $slug ] ) ) {
+                    // Skip recursive patterns.
+                    array_splice( $blocks, $i, 1 );
+                    continue;
+                }
+
+                $registry = WP_Block_Patterns_Registry::get_instance();
+                $pattern  = $registry->get_registered( $slug );
+
+                // Skip unknown patterns.
+                if ( ! $pattern ) {
+                    ++$i;
+                    continue;
+                }
+                $blocks_to_insert = parse_blocks( trim( $pattern['content'] ) );
+
+                // For single-root patterns, add the pattern name to make this a pattern instance in the editor.
+                if ( count( $blocks_to_insert ) === 1 ) {
+                    $metadata = array(
+                        'patternName' => $slug,
+                    );
+
+                    if ( ! empty( $pattern['title'] ) ) {
+                        $metadata['name'] = sanitize_text_field( $pattern['title'] );
+                    }
+
+                    if ( ! empty( $pattern['description'] ) ) {
+                        $metadata['description'] = sanitize_text_field( $pattern['description'] );
+                    }
+
+                    if ( ! empty( $pattern['categories'] ) && is_array( $pattern['categories'] ) ) {
+                        $metadata['categories'] = array_map( 'sanitize_text_field', $pattern['categories'] );
+                    }
+
+                    $blocks_to_insert[0]['attrs']['metadata'] = $metadata;
+                }
```

Comment:
These are the main changes from https://github.com/WordPress/wordpress-develop/pull/10248